### PR TITLE
Lab/NB7 extension compatibility docs

### PR DESCRIPTION
**Pull request by ericsnekbytes**
Adds a new extension-compatibility document to the documentation. Some sample code will also be committed to the examples repo, related PRs:

- Top area text widget
  - https://github.com/jupyterlab/extension-examples/pull/240
- Shout button with optional extra Lab features
  - https://github.com/jupyterlab/extension-examples/pull/243
- Clap button
  - https://github.com/jupyterlab/extension-examples/pull/244
- ~~Step counter (provider-consumer example in 3 related packages)~~ moved to #15417

### CODE REVIEW

File: docs/source/extension/extension_dual_compatibility.rst
Comment (jtpio) [inline docs/source/extension/extension_dual_compatibility.rst:4]: Since this would apply to more lab-based applications than just JupyterLab and Notebook 7, maybe something like the following would work?

```suggestion
Building extensions for JupyterLab-based applications
```

Comment (jtpio) [reply docs/source/extension/extension_dual_compatibility.rst:4]: And then have a section focused on JupyterLab 4 and Notebook 7 since they are the two major notebook frontends at the moment built with this stack.

File: docs/source/extension/extension_dual_compatibility.rst
Comment (jtpio) [inline docs/source/extension/extension_dual_compatibility.rst:31]: To disambiguate and for consistency we'll probably want to stick to "JupyterLab" instead of "Lab" in this document.

```suggestion
blocks and methods. For instance, both JupyterLab and Notebook 7 accept Lumino widgets
```

File: docs/source/extension/extension_dual_compatibility.rst
Comment (jtpio) [inline docs/source/extension/extension_dual_compatibility.rst:87]: Technically there is a `main` area in Notebook 7. But only one widget can be added to it.

We should update that sentence to reflect that.

File: docs/source/extension/extension_dual_compatibility.rst
Comment (jtpio) [inline docs/source/extension/extension_dual_compatibility.rst:98]: It would be better to not mention `INotebookShell` to prevent extensions authors from relying too much on it.

Also `INotebookShell` does not provide much than just the `currentChanged` signal compared to a `JupyterFront.IShell`.

Comment (jtpio) [reply docs/source/extension/extension_dual_compatibility.rst:98]: Instead we can mention how widgets can be positioned in other areas than the `main` area in both JupyterLab and Notebook 7:

- https://jupyterlab.readthedocs.io/en/latest/user/interface_customization.html
- https://jupyter-notebook.readthedocs.io/en/latest/configuring/interface_customization.html

And expand the documentation there if needed.

File: docs/source/extension/extension_dual_compatibility.rst
Comment (ericsnekbytes) [reply docs/source/extension/extension_dual_compatibility.rst:6]: This idea is going to be hard to communicate to people: NotebookFrontend is used to build Lab and NB7, other frontends may come out in the future, and that you can aim for compatibility with all of them. I think it would be easier to discuss Lab/NB7 first and then have a section later discussing other frontends that the other way around (kind of the reverse of what you're suggesting) plus a bit of minor rewording to accommodate these new concepts. Introducing all of these concepts up front I think is too heavy at the beginning of this document.

Comment (jtpio) [reply docs/source/extension/extension_dual_compatibility.rst]: > NotebookFrontend is used to build Lab and NB7

You mean `JupyterFrontEnd`?

Many extensions that provide commands, themes, menu entries, notebook toolbar buttons... don't need to care about the shell of the host application. Extensions should focus on Extension Points as defined in https://jupyterlab.readthedocs.io/en/latest/extension/extension_points.html.

Some applications might or might not provide some of these extension points, as it is the case with Notebook 7.

Trying to convey this idea in this document would be great so extension authors can follow best practices from the beginning. And would be coherent with the existing documentation.

Comment (ericsnekbytes) [reply docs/source/extension/extension_dual_compatibility.rst]: > You mean `JupyterFrontEnd`?

Yeah I meant this.

> Many extensions that provide commands, themes, menu entries, notebook toolbar buttons... don't need to care about the shell of the host application. Extensions should focus on Extension Points as defined in https://jupyterlab.readthedocs.io/en/latest/extension/extension_points.html.

I will definitely incorporate this, thanks for the feedback. I would probably call this the provider/consumer pattern here (I think this is how it is discussed in other parts of the docs already).

I still think it's useful to tell people how they can tailor specialized behaviors for a specific frontend by testing the kernel and using some thin frontend-specific UI code (and this is one of the approaches we discussed in the Notebook meeting). Though I can add some warnings/caveats about overusing this approach and how the other pattern should be the default.

> Some applications might or might not provide some of these extension points, as it is the case with Notebook 7.
> 
> Trying to convey this idea in this document would be great so extension authors can follow best practices from the beginning. And would be coherent with the existing documentation.

Will do!

File: docs/source/extension/extension_dual_compatibility.rst
Comment (ericsnekbytes) [reply docs/source/extension/extension_dual_compatibility.rst]: I left some notes about INotebookShell and tailoring frontend-specific behaviors above (need to give some warnings but I think still valid in summary).

Other areas I can definitely discuss.

File: docs/source/extension/extension_dual_compatibility.rst
Comment (jtpio) [reply docs/source/extension/extension_dual_compatibility.rst]: > notification area too

The notification plugin has been added to Notebook 7: https://github.com/jupyter/notebook/pull/6900

But I'm not sure the notification popups are displayed correctly at the moment if an extension uses the notification API. But in theory they should work, at least partially.

File: docs/source/extension/extension_dual_compatibility.rst
Comment (ericsnekbytes) [reply docs/source/extension/extension_dual_compatibility.rst]: These should be resolved now.

File: docs/source/extension/extension_dual_compatibility.rst
Comment (ericsnekbytes) [reply docs/source/extension/extension_dual_compatibility.rst]: This section was updated, should be resolved.

File: docs/source/extension/plugin_system.rst
Comment (jtpio) [inline docs/source/extension/plugin_system.rst:4]: Wondering if this deserves a new page or if it could go in https://jupyterlab.readthedocs.io/en/latest/extension/extension_dev.html ?

Comment (jtpio) [reply docs/source/extension/plugin_system.rst:4]: Or maybe we can move the current content of https://jupyterlab.readthedocs.io/en/latest/extension/extension_dev.html to this new `plugin_system.rst`, so it shows up in the side bar as "The JupyterLab Plugin System":

image

Comment (ericsnekbytes) [reply docs/source/extension/plugin_system.rst:4]: This is a good question. I started this document because I thought there was a gap between the very high level explanation offered in extension_dev and the technical steps and understanding someone needs to use it in the code. Maybe it should be integrated and the page renamed, which could be beneficial...though it would take some refactoring of the main document for sure. I'll take a look.

Comment (ericsnekbytes) [reply docs/source/extension/plugin_system.rst:4]: I decided to keep this document separate so that it functions as a focused high level design document (with an example). The main page can use some refactoring, maybe as a future PR...

**Comment by ericsnekbytes** [comment]
@jtpio The last example (the provider-consumer pattern demo) is nearly complete (minus a trivial addition of an update signal for the step count and some minor renaming) if you want to take a look at that PR (more info there). I'll be out for ~2 weeks, and will resume then, though it's nearing completion as-is (the prior examples haven't been cleaned and documented as well as the provider-consumer example yet).

**Comment by ericsnekbytes** [comment]
@jtpio If you have time to review I'd like to get this approved, then merge the extension PRs FIRST before this one. Thanks :)

File: docs/source/extension/extension_dev.rst
Comment (jtpio) [inline docs/source/extension/extension_dev.rst:50]: Should this new page be added to the sidebar, so it's easier to find?

image

File: docs/source/extension/plugin_system.rst
Comment (krassowski) [inline docs/source/extension/plugin_system.rst:4]: Could this be possibly moved out into a separate pull request? Two reasons:
1. We could more quickly merge the dual compatibility doc delivering the excellent piece above to the users/developers
2. This document would receive more attention from a wider group of developers, because there are folks who review pull request based on the title, I frankly did not expect to find this document in a PR titled ` Open
Lab/NB7 extension compatibility docs` :)

File: docs/source/extension/plugin_system.rst
Comment (krassowski) [inline docs/source/extension/plugin_system.rst:179]: If we are creating a separate doc, it would be good to cover
- `deactivate()`
- mention naming conventions for identifiers
- check if we need to expose anything else from lumino's `IPlugin`
- deferred plugins
- maybe `async activate` and what it implies

File: docs/source/extension/plugin_system.rst
Comment (krassowski) [inline docs/source/extension/plugin_system.rst:11]: I appreciate the need for naming things, but I wonder if this does not make the document less approachable. This section has a substantial overlap with [Plugins Interacting with Each Other](https://github.com/jupyterlab/jupyterlab/blob/main/docs/source/extension/extension_dev.rst#plugins-interacting-with-each-other) which is much friendlier title for a section in my opinion.

File: docs/source/extension/extension_dev.rst
Comment (ericsnekbytes) [reply docs/source/extension/extension_dev.rst:50]: It should, do you know how? I've been poking around a bit but haven't found out yet...

File: docs/source/extension/plugin_system.rst
Comment (ericsnekbytes) [reply docs/source/extension/plugin_system.rst:179]: I can add these, probably in a separate PR 👍

File: docs/source/extension/plugin_system.rst
Comment (ericsnekbytes) [reply docs/source/extension/plugin_system.rst:11]: I agree that some renaming/a friendlier title would be good here, though still think it's important to keep the term in there so people have something to point toward that describes the paradigm consicely.

**Comment by ericsnekbytes** [comment]
I split off the plugin_system and provider/consumer pattern document into[ a new PR](https://github.com/jupyterlab/jupyterlab/pull/15417).

File: docs/source/extension/plugin_system.rst
Comment (ericsnekbytes) [reply docs/source/extension/plugin_system.rst:4]: The plugin system doc has been split out [here](https://github.com/jupyterlab/jupyterlab/pull/15417).

File: docs/source/extension/plugin_system.rst
Comment (ericsnekbytes) [reply docs/source/extension/plugin_system.rst:179]: The plugin system doc has been split out [here](https://github.com/jupyterlab/jupyterlab/pull/15417).

File: docs/source/extension/extension_dev.rst
Comment (ericsnekbytes) [reply docs/source/extension/extension_dev.rst]: Okay, I got it working. For others: I just needed to add the page name (filename no ext.) to the toctree directive for the extension_dev page (near the top extension_dev.rst):

```
.. toctree::
   :maxdepth: 1

   extension_points
   ui_components
   documents
   notebook
   virtualdom
   ui_helpers
   internationalization
   identity
   extension_tutorial
   extension_dual_compatibility
   extension_migration
```

File: docs/source/extension/plugin_system.rst
Comment (jtpio) [reply docs/source/extension/plugin_system.rst:11]: > substantial overlap with Plugins Interacting with Each Other

Also wondering why we need to re-explain the concepts already covered in [Plugins Interacting with Each Other](https://github.com/jupyterlab/jupyterlab/blob/main/docs/source/extension/extension_dev.rst#plugins-interacting-with-each-other)? That existing section is maybe a bit longer than the new one proposed in this PR, but still concise.

File: docs/source/extension/plugin_system.rst
Comment (jtpio) [inline docs/source/extension/plugin_system.rst:69]: Maybe use "JupyterLab" instead of "Jupyter"? So it's consistent with the rest of the docs in https://jupyterlab.readthedocs.io/en/latest/extension/extension_dev.html#plugins?

File: docs/source/extension/extension_dev.rst
Comment (jtpio) [inline docs/source/extension/extension_dev.rst:49]: Why not call the document "Make an extension for Jupyter Notebook 7" then?

Comment (jtpio) [reply docs/source/extension/extension_dev.rst:49]: It's not clear at first glance what "Dual-Compatible" means when just looking at the title.

File: docs/source/extension/plugin_system.rst
Comment (ericsnekbytes) [reply docs/source/extension/plugin_system.rst:69]: This doc was moved to a new PR here, so I added your suggestion there.

File: docs/source/extension/extension_dev.rst
Comment (ericsnekbytes) [reply docs/source/extension/extension_dev.rst]: I'll switch to "Making Extensions Compatible with Multiple Applications Tutorial" to emphasize the goal, which is compatibility (@krassowski suggested something similar with the provider-consumer examples, maybe it's better to focus on the end goal).

Putting both "JupyterLab" and "Jupyter Notebook 7" makes the title excessively long, and adding the concept of Frontends in the title will probably add confusion more than clarify, as its meaning in this context will likely be unknown to those outside the circle of core devs.

**Comment by ericsnekbytes** [comment]
@fcollonval I forgot to tag you!

**Comment by krassowski** [comment]
>  think if our goal is to include chemists, physicists, other technical non-programmers
> ...
>  "Dependency Injection Pattern" is actively harmful to that goal. Using "Provider-Consumer Pattern"

One is not accessible to non-dev native speakers, the other is not accessible to ESL speakers ;)

Do we need to name it a pattern? Could it be just "Interacting with other plugins", "plugins-interactions" or similar?

**Comment by ericsnekbytes** [comment]
> > think if our goal is to include chemists, physicists, other technical non-programmers
> > ...
> > "Dependency Injection Pattern" is actively harmful to that goal. Using "Provider-Consumer Pattern"
> 
> One is not accessible to non-dev native speakers, the other is not accessible to ESL speakers ;)
> 
> Do we need to name it a pattern? Could it be just "Interacting with other plugins", "plugins-interactions" or similar?

I think we can serve both audiences by using plain language up front, and linking to the expanded explanation where it is used in the text. One problem we've had with docs is trying to cram too much information into one place or article, and with this approach a reader experiences it like a Wikipedia article, where they can jump into deep links when they feel they need more explanation. To summarize the benefits:

- Benefits of using "The provider-consumer pattern (a type of dependency injection pattern)":
  - "provider-consumer" gives a concise name to this complex "requires" + "optional" + "activate args" design/implementation
  - Uses more friendly terminology that more accurately labels what is being done
  - Also describes it in more detail, using terminology familiar to software veterans, either in parentheses or in a deep link to an expanded explanation (which right now is a section on the "extension_dev" page)
  - By deep linking, we relieve the burden of trying to cram all of this complex and nuanced information inside of our article, and give the reader an option to pursue it further at their own pace

**Comment by ericsnekbytes** [comment]
I pushed some edits/refactors as discussed above, and verified that the code snippets are up to date with the current version of the examples repo. Waiting for CI now.

File: docs/source/extension/extension_multiple_ui.rst
Comment (krassowski) [inline docs/source/extension/extension_multiple_ui.rst:9]:
```suggestion
*Make your extensions work in JupyterLab, Notebook 7+ and more!*
```

File: docs/source/extension/extension_multiple_ui.rst
Comment (krassowski) [inline docs/source/extension/extension_multiple_ui.rst:29]: This paragraph does not answer the question yet

```suggestion
Getting started
---------------

At a high level, extensions for JupyterLab and Jupyter Notebook both
typically start from a template project, which you can download and set up
using instructions from `the extension tutorial <https://jupyterlab.readthedocs.io/en/latest/extension/extension_tutorial.html>`_).
Once your template is ready, you can start adding components and features to build your extension.

How Compatibility Works
-----------------------

```

Comment (krassowski) [reply docs/source/extension/extension_multiple_ui.rst:29]: + we should use sphinx references here, not fully specified links

File: docs/source/extension/extension_multiple_ui.rst
Comment (krassowski) [inline docs/source/extension/extension_multiple_ui.rst:30]: Should use sphinx reference, not a fully specified link

File: docs/source/extension/extension_multiple_ui.rst
Comment (krassowski) [inline docs/source/extension/extension_multiple_ui.rst:35]: Both is ambig

```suggestion
(among others) to help build your extension's look and behavior. Both
Lumino and JupyterLab API are written in Typescript.
```

File: docs/source/extension/extension_multiple_ui.rst
Comment (krassowski) [inline docs/source/extension/extension_multiple_ui.rst:46]:
```suggestion
Compatibility can be achieved without additional effort in some simple
```

or with "little effort". "literally no effort" will be frustrating when it does not work for the user (see https://jupyterlab.readthedocs.io/en/latest/developer/contributing.html#writing-style)

File: docs/source/extension/extension_multiple_ui.rst
Comment (krassowski) [inline docs/source/extension/extension_multiple_ui.rst:122]: Please change to a shpinx reference instead of fully specified link

File: docs/source/extension/extension_multiple_ui.rst
Comment (krassowski) [inline docs/source/extension/extension_multiple_ui.rst:39]:
```suggestion
as interface components, both apps enable you to add them to the interface by
```

File: docs/source/extension/extension_multiple_ui.rst
Comment (krassowski) [inline docs/source/extension/extension_multiple_ui.rst:7]: This is kind of long. I  think we can remove "Extensions" because this doc is in "Develop Extensions" chapter.

```suggestion
Targeting Multiple Applications
===============================
```

Do you feel strongly about keeping "Extensions" part in?

image

File: docs/source/extension/extension_multiple_ui.rst
Comment (krassowski) [inline docs/source/extension/extension_multiple_ui.rst:223]: I would keep it only for the first time (above) to reduce noise - this tutorial is not about widgets.

```suggestion
        clapWidget.id = 'JupyterLabClapWidgetLab';
```

File: docs/source/extension/extension_multiple_ui.rst
Comment (krassowski) [inline docs/source/extension/extension_multiple_ui.rst:243]: Same as above

```suggestion
        clapWidget.id = 'JupyterNotebookClapWidgetNotebook';
```

File: docs/source/extension/extension_multiple_ui.rst
Comment (krassowski) [inline docs/source/extension/extension_multiple_ui.rst:209]: This code snippets do not read great on more narrow screens, with some strings and comments requiring scrolling. Here is one case that might be easy cut:

```suggestion
     * Data for the @jupyterlab-examples/clap-button JupyterLab plugin.
```

Note `plugin` is more correct here than `extension` in technical sense.

File: docs/source/extension/extension_multiple_ui.rst
Comment (krassowski) [inline docs/source/extension/extension_multiple_ui.rst:229]:
```suggestion
     * Data for the @jupyterlab-examples/clap-button Jupyter Notebook plugin.
```

**Review by krassowski — COMMENTED** [review]
Thank you! Some small suggestions before we merge it

**Comment by ericsnekbytes** [comment]
I changed the title, and made another PR for some of the extra suggestions. Waiting for CI build to review.

**Comment by krassowski** [comment]
There is a failure in the docs test, could you take a look?

image

**Comment by ericsnekbytes** [comment]
> There is a failure in the docs test, could you take a look?
> 
> image

Resolved 👍

**Review by krassowski — APPROVED** [review]
Thanks!

**Pull request merged**
